Xilinx-CNS/kubernetes-onload
# ON-15119: Add device plugin

~~Still in progress, there is some code copied from #47 which will have to removed once that is merged~~ Done
### Patch 1/3
Adds support for the onload device plugin to the controller using existing image from `v3.0.0-pre`.
### Patch 2/3
Adds the source code for the device plugin from `v3.0.0-pre` with some minor modifications so it can handle being in a different directory structure.
Also added new commands to the makefile to build and push the device plugin image
### Patch 3/3
Updates the onload controller to use the images produced by patch 2
### Patch 4/3
Separates the device plugin into `pkg/deviceplugin` and `cmd/deviceplugin` to better align with the [golang recommended project layout](https://github.com/golang-standards/project-layout).
 - `pkg/deviceplugin` still contains the main logic for the device plugin, but now in a package called `deviceplugin`
 - `cmd/deviceplugin` contains the `main` function for the device plugin and acts as an entrypoint into the package.
## Testing done

#### Automated test
```console
$ make test
KUBEBUILDER_ASSETS="/home/tcrawley/Dev/kubernetes-onload/bin/k8s/1.26.0-linux-amd64" go test ./... -coverprofile cover.out
?       github.com/Xilinx-CNS/kubernetes-onload [no test files]
?       github.com/Xilinx-CNS/kubernetes-onload/api/v1alpha1    [no test files]
?       github.com/Xilinx-CNS/kubernetes-onload/pkg/deviceplugin        [no test files]
ok      github.com/Xilinx-CNS/kubernetes-onload/controllers     8.221s  coverage: 77.3% of statements
```

#### Manual testing
 * Built both device plugin image and controller image and pushed to dev registry.
 * Deployed operator using `make deploy`
 * Added Onload CR using `oc create -k config/samples` & verified all Modules/DaemonSets were created
 * Tested device plugin
   * Created an example pod requesting an `onload` resource -- succeeds with the device plugin active and fails otherwise
   * Deployed sfnettest pods from `v3.0.0-pre` and verified accelerated traffic was passing
  * Deleted Onload CR (after deleting user pods) to verify everything was removed
